core/network/src/protocol.rs

```diff
@@ -66,9 +67,27 @@ pub struct Protocol<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> {
 	consensus_gossip: ConsensusGossip<B>,
 	context_data: ContextData<B, H>,
 	// Connected peers pending Status message.
-	handshaking_peers: HashMap<NodeIndex, time::Instant>,
+	handshaking_peers: HashMap<NodeIndex, HandshakingPeer>,
+	// Connected peers from whom we received a Status message,
+	// similar to context_data.peers but shared with the SyncProvider.
+	connected_peers: Arc<RwLock<HashMap<NodeIndex, ConnectedPeer<B>>>>,
 	transaction_pool: Arc<TransactionPool<H, B>>,
 }
+
+#[derive(Clone)]
+/// A peer from whom we have received a Status message.
+pub struct ConnectedPeer<B: BlockT> {
+	pub peer_id: Option<PeerId>,
```

Comment:
This should not be an option, since we go from handshake -> connected and we already know the peer id on the handshake. I guess we can even remove this field if it's also not defined as `Option` in `PeerInfo`.